File: src/org/virtue/game/entity/combat/impl/Prayer.java
/**
 * Copyright (c) 2014 Virtue Studios
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package org.virtue.game.entity.combat.impl;

import org.virtue.config.structtype.StructType;
import org.virtue.game.entity.Entity;
import org.virtue.game.entity.player.Player;

/**
 * @author Im Frizzy <skype:kfriz1998>
 * @author Frosty Teh Snowman <skype:travis.mccorkle>
 * @author Arthur <skype:arthur.behesnilian>
 * @author Sundays211
 * @since Feb 6, 2015
 */
public class Prayer {

	/**
	 * Type
	 */
	private StructType type;
	
	/**
	 * StructType
	 * @return
	 */
	public StructType getType () {
		return type;
	}
	
	/**
	 * The {@link Entity} for this {@link Prayer}
	 */
	private Player player;
	
	/**
	 * Using Prayer True/False
	 */
	public boolean usingPrayer = false;
	
	/**
	 * The amount of prayer points this {@link Entity} has
	 */
	private int prayer = -1;
	
	private int prayerEnergy = 0;
	
	private int[] spriteIDS = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 7, 2, 1, 0, 3, 5, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
	
	double[] prayerDrainRate = { 1.2, 1.2, 1.2, 1.2, 1.2, 0.6, 0.6, 0.6, 3.6, 1.8, 1.8, 0.6, 0.6, 0.3, 0.3, 0.3, 0.3, 0.3, 0.3, 0.3, 0.3, 0.3, 1.2, 0.6, 0.18, 0.18, 0.24, 0.15, 0.2, 0.18 };
	
	/**
	 * @param player
	 */
	public Prayer(Player player) {
		this.player = player;
	}
	
	/**
	 * Activates Prayer
	 * @param slot
	 */
	public void activate(int slot) {
		player.getHeadIcons().setIcon(440, spriteIDS[slot]);
		player.getHeadIcons().refresh();
		usingPrayer = true;
	}
	
	/**
	 * De-activates Prayer
	 * @param slot
	 */
	public void deactivate(int slot) {
		player.getHeadIcons().reset();
		player.getHeadIcons().refresh();
		usingPrayer = false;
	}
	
	/**
	 * @return the prayer
	 */
	public int getPrayerPoints() {
		return prayer;
	}

	/**
	 * @param prayer the prayer to set
	 */
	public void setPrayerPoints(int prayer) {
		if ((this.prayer - prayer < 0) && (this.prayer != -1))
			prayer = 0;
		
		this.prayer = prayer;
		
		if (player instanceof Player) {
			((Player) player).getDispatcher().sendPrayer(this.prayer * 10);
		}
	}
	
	public boolean drainPrayer() {//TODO Prayer Draining
		float drain = getPrayerPoints();
		
		prayerEnergy -= drain;
		if (prayerEnergy <= 0) {
			//usingPrayer = false;
			//setPrayerPoints(0);
			return false;
		} else {
			usingPrayer = true;
			player.getDispatcher().sendPrayer((int) prayerEnergy);
			return true;
		}
	}
}

File: repository/scripts/skill/Prayer.js
/**
 * Copyright (c) 2014 Virtue Studios
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
 * @author Kayla
 * @since 19/11/2014
 */

var Bury = {
		IMPIOUS_ASHES : {
			itemID : 20264,
			xp : 4
		},
		BONES : {
			itemID : 526,
			xp : 4.5
		},
		WOLF_BONES : {
			itemID : 2859,
			xp : 4.5
		},
		BURNT_BONES : {
			itemID : 528,
			xp : 4.5
		},
		MONKEY_BONES : {
			itemID : 3179,
			xp : 5
		},
		BAT_BONES : {
			itemID : 17672,
			xp : 5.3
		},
		ACCURSED_ASHES : {
			itemID : 20266,
			xp : 12.5
		},
		BIG_BONES : {
			itemID : 17674,
			xp : 12.5
		},
		JOGRE_BONES : {
			itemID : 3125,
			xp : 15
		},
		ZOGRE_BONES : {
			itemID : 4812,
			xp : 22.5
		},
		SHAIKAHAN_BONES : {
			itemID : 3123,
			xp : 22.5
		},
		BABYDRAGON_BONES : {
			itemID : 534,
			xp : 30
		},
		WYVERN_BONES : {
			itemID : 6812,
			xp : 50
		},
		INFERNAL_ASHES : {
			itemID : 20268,
			xp : 62.5
		},
		DRAGON_BONES : {
			itemID : 17676,
			xp : 72
		},
		FAYRG_BONES : {
			itemID : 4830,
			xp : 84
		},
		RAURG_BONES : {
			itemID : 4832,
			xp : 96
		},
		DAGANNOTH_BONES : {
			itemID : 6729,
			xp : 125
		},
		AIRUT_BONES : {
			itemID : 30209,
			xp : 132.5
		},
		OURG_BONES : {
			itemID : 4834,
			xp : 140
		},
		FROST_DRAGON_BONES : {
			itemID : 18830,
			xp : 180
		},
		ANCIENT_BONES : {
			itemID : 15410,
			xp : 200
		}
	};
	
var PrayerListener = Java.extend(Java.type('org.virtue.engine.script.listeners.EventListener'), {
	invoke : function (event, objTypeId, args) {
		buryBones(args.player, objTypeId, args.slot);
	}
});

/* Listen to the item ids specified */
var listen = function(scriptManager) {
	var ids = [];
	for (var i in Bury) {
		ids.push(Bury[i].itemID);
	}
	var listener = new PrayerListener();
	for (var i in ids) {
		//Bind option one on all bones/ashes to this listener
		scriptManager.registerListener(EventType.OPHELD1, ids[i], listener);
	}	
}

var Prayer = {
		buryBones : function (player, objTypeId, slot) {
			if (api.isPaused(player)) {
				return false;//Cannot bury as an action is already in process
			}
			var bones = this.forBones(player, objTypeId);
			if(bones == null) {
				return true;
			}
			var delay = 3;
			api.pausePlayer(player, delay+1);
			api.sendMessage(player, "You dig a hole in the ground...", MesType.GAME_SPAM);
			api.runAnimation(player, 827);
			api.delCarriedItem(player, objTypeId, 1, slot);
			var Action = Java.extend(Java.type('org.virtue.game.entity.player.event.PlayerActionHandler'), {	
				process : function (player) {
					api.runAnimation(player, 827);
					if (delay <= 0) {
						Prayer.burySuccess(player, bones);
						return true;
					}
					delay--;
					return false;
				},
				stop : function (player) {//Clear the current animation block
					api.stopAnimation(player);
				}
			
			});
			player.setAction(new Action());	
		},
		forBones : function (player, itemId) {
			var bones;
			for (var ordial in Bury) {
				bones = Bury[ordial];
				if (bones.itemID == itemId) {
					return bones;
				}
			}
			return null;
		},
		burySuccess : function (player, bones) {
			api.sendMessage(player, "You bury the " + api.getItemName(bones.itemID) + ".", MesType.GAME_SPAM);
			api.addExperience(player, Stat.PRAYER, bones.xp, true);
		}
}
